File: LinkedList/LLCycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
 
 // Edge cases: [], [1,2] in the while loop cuz no fast.next.next
public class Solution {
    public boolean hasCycle(ListNode head) {
    if (head == null) return false; // If empty LL then no cycles
    ListNode fast = head, slow = head; // Two ref pointers, fast and slow
    while (fast != null && fast.next != null) { // Make sure fast doesn't fall off
        slow = slow.next; // slow goes by one
        fast = fast.next.next; // fast moves by two
        if (slow == fast) return true; // if they both catch up ==> CYCLE DETECTED
    }
    return false; // finished loop so no cycles
    }
}

File: LinkedList/LLCycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a boolean
    def hasCycle(self, head):
        if head == None:
            return False
        fast = head
        slow = head
        while (fast and fast.next): # fast goes before fast.next
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False
